app/grandchallenge/algorithms/tasks.py

```diff
@@ -163,105 +160,134 @@ def create_algorithm_jobs(
+def filter_archive_items_for_algorithm(
+    *, archive_items, algorithm_image, algorithm_model
+):
     """
-    Removes sets of civs that are invalid for new jobs
+    Removes archive items that are invalid for new jobs.
+    The archive items need to contain values for all inputs of one of the algorithm's interfaces.
 
     Parameters
     ----------
-    civ_sets
-        Iterable of sets of ComponentInterfaceValues that are candidate for
-        new Jobs
+    archive_items
+        Archive items whose values are candidates for new jobs' inputs
     algorithm_image
         The algorithm image to use for new job
     algorithm_model
         The algorithm model to use for the new job or None
 
     Returns
     -------
-    Filtered set of ComponentInterfaceValues
+    Dictionary of valid ArchiveItems for new jobs, grouped by AlgorithmInterface
     """
     from grandchallenge.algorithms.models import Job
+    from grandchallenge.evaluation.models import (
+        get_archive_items_for_interfaces,
+    )
 
-    input_interfaces = {*algorithm_image.algorithm.inputs.all()}
+    algorithm_interfaces = (
+        algorithm_image.algorithm.interfaces.prefetch_related("inputs").all()
+    )
 
+    # First, sort archive items by algorithm interface:
+    # An archive item is only valid for an interface if it has values
+    # for all inputs of the interface
+    valid_job_inputs = get_archive_items_for_interfaces(
+        algorithm_interfaces=algorithm_interfaces, archive_items=archive_items
+    )
+    valid_input_counts = [
+        interface.inputs.count() for interface in algorithm_interfaces
+    ]
+
+    # Next, get all system jobs that have been run for the provided archive items
+    # with the same model and image
+    if algorithm_model:
+        extra_filter = {"algorithm_model": algorithm_model}
+    else:
+        extra_filter = {"algorithm_model__isnull": True}
     existing_jobs = {
         frozenset(j.inputs.all())
         for j in Job.objects.filter(
-            algorithm_image=algorithm_image, algorithm_model=algorithm_model
+            algorithm_image=algorithm_image,
+            algorithm_interface__in=valid_job_inputs.keys(),
+            creator=None,
+            **extra_filter,
         )
         .annotate(
+            input_count=Count("inputs", distinct=True),
             inputs_match_count=Count(
                 "inputs",
                 filter=Q(
-                    inputs__in={civ for civ_set in civ_sets for civ in civ_set}
+                    inputs__in={
+                        civ
+                        for archive_items in valid_job_inputs.values()
+                        for item in archive_items
+                        for civ in item.values.all()
+                    }
                 ),
-            )
+                distinct=True,
+            ),
+        )
+        .filter(
+            input_count=F("inputs_match_count"),
+            input_count__in=valid_input_counts,
```

Comment:
I have a feeling this isn't quite right, don't you need to limit it to the count of the number of inputs for the one relevant interface, not all interfaces? So you would then need to find:

- What interface does this job match?
- How many inputs are there for this interface?
- Does the number of inputs match the number of inputs for this interface?

You could do this slowly by taking the original query and then iterating over each interface.

Reply:
Hm, you might be right. It's similar to the `successful_jobs` property further below.

Reply:
Regarding "What interface does this job match?" - the job has an interface set, but you're suggesting that we should not trust the a job is actually valid for the interface that is set on the job, and hence double check here? 

I think that is indeed necessary given how we decided to migrate the interfaces for existing jobs. I think then I should first subset to jobs that have inputs that conform with their configured interfaces, and only then do the input comparison with the archive items.

Reply:
I have updated the logic here and in the model property. They now use the same underlying code to get valid jobs per interface